lib/wasi.js

```diff
@@ -92,9 +89,19 @@ class WASI {
         'instance.exports._initialize', 'undefined', _initialize);
     }
 
-    if (!(memory instanceof WebAssembly.Memory)) {
+    // WASI::_SetMemory() in src/node_wasi.cc only expects that |memory| is
+    // an object. It will try to look up the .buffer property when needed
+    // and fail with UVWASI_EINVAL when the property is missing or is not
+    // an ArrayBuffer. Long story short, we don't need much validation here
+    // but we type-check anyway because it helps catch bugs in the user's
+    // code early.
+    validateObject(memory, 'instance.exports.memory');
```

Comment:
non-blocking observation: one additional check we could do here is to check that `memory.constructor.name === 'Memory'`.

Reply:
I don’t see why anybody would ever subclass `WebAssembly.Memory`, but then again I don’t really see why we would ever want to forbid that. I think this can stay as-is.